src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java

```diff
@@ -0,0 +1,138 @@
+                             PermissionsService permissionService,
+                             SafeDepositBoxService safeDepositBoxService) {
+
+        super(secureDataService, permissionService, safeDepositBoxService);
+    }
+
+    @Override
+    protected ResponseInfo<Object> generateVaultStyleResponse(VaultStyleErrorResponse response, int statusCode) {
+        return ResponseInfo.newBuilder()
+                .withContentForFullResponse(response)
+                .withHttpStatusCode(statusCode)
+                .build();
+    }
+
+    @Override
+    public CompletableFuture<ResponseInfo<Object>> doExecute(SecureDataRequestInfo requestInfo,
+                                                                         RequestInfo<Void> request,
+                                                                         Executor longRunningTaskExecutor,
+                                                                         ChannelHandlerContext ctx,
+                                                                         SecurityContext securityContext) {
+
+        if (StringUtils.equalsIgnoreCase(request.getQueryParamSingle("list"), "true")) {
+            return CompletableFuture.supplyAsync(
+                    AsyncNettyHelper.supplierWithTracingAndMdc(() -> listKeys(requestInfo), ctx),
+                    longRunningTaskExecutor
+            );
+        }
+
+        return CompletableFuture.supplyAsync(
+                AsyncNettyHelper.supplierWithTracingAndMdc(() -> readSecureData(requestInfo), ctx),
+                longRunningTaskExecutor
+        );
+    }
+
+    private ResponseInfo<Object> listKeys(SecureDataRequestInfo info) {
+        Set<String> keys = secureDataService.listKeys(info.getPath());
+
+        if (keys.isEmpty()) {
+            return generateVaultStyleResponse(new VaultStyleErrorResponse.VaultStyleErrorResponseBuilder().build(),
+                    HttpResponseStatus.NOT_FOUND.code());
+        }
+
+        SecureDataResponse response = new SecureDataResponse();
+        response.setRequestId(UUID.randomUUID().toString()); // maybe use trace id?
+        response.setData(ImmutableMap.of("keys", keys));
+
+        return ResponseInfo.newBuilder()
+                .withContentForFullResponse(response)
+                .withHttpStatusCode(HttpResponseStatus.OK.code())
+                .build();
+    }
+    private ResponseInfo<Object> readSecureData(SecureDataRequestInfo info) {
+
+        Optional<String> data = secureDataService.readSecret(info.getPath());
+
+        if (! data.isPresent()) {
+            return generateVaultStyleResponse(new VaultStyleErrorResponse.VaultStyleErrorResponseBuilder().build(),
+                    HttpResponseStatus.NOT_FOUND.code());
+        }
+
+        SecureDataResponse response = new SecureDataResponse();
+        response.setRequestId(UUID.randomUUID().toString());
+        try {
+            response.setData(new ObjectMapper().readTree(data.get()));
```

Comment:
Are we converting to JSON here?  Wasn't there one SDB where they had used Vault APIs directly and didn't do the normal key/value pairs?  Maybe we need the internals to support arbitrary files from day one?

Reply:
Yes this is taking the object structure and loading into a JsonNode, because the Object can be any valid json you can not cast it to a Map<String, String> it must be treated as generic JSON aka a JsonNode.

I manually tested it, and it works.